[inductor] Constant folding for dynamic shape node before pattern matching

imzhuhl opened the pull request:
Extend constant folding for dynamic shape node, only support pointwise op and some restricted ops 

We support dynamic shapes by limiting constant folding of ops that are guaranteed to have uniform values (`full`, pointwise ops, and views) and running these operators with tensors of shape 1. This also eliminates the possibility of memory overhead of constant folding.

Test plan:

```
python test/inductor/test_torchinductor_dynamic_shapes.py -k test_constant_fold_uniform_value_dynamic
```

Fixes #128387

@eellison 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @ezyang @bdhirsh @anijain2305

imzhuhl committed:
[inductor] Constant folding for dynamic shape node before pattern matching

Extend constant folding for dynamic shape node, only support pointwise op and some restricted ops

Test plan:

```
python test/inductor/test_torchinductor_dynamic_shapes.py -k test_constant_fold_uniform_value_dynamic
```

eellison reviewed (commented):
looking good ! we just got a request to reduce memory usage from Uniform Value constant folder. i know this is not exactly what was originally scoped but it would be great to fold the two changes together since I think they are similar.

I'm thinking along the lines of:

- add a mechanism for custom running of node in ConstantFold

- Within ConstantFoldUniformValue - For non constructors, non-pointwise nodes, non-views return `unknown_value` .

- Otherwise:

For constructors, always construct a tensor of size [1]
For views, just return the input
Pointwise nodes can just defer to the normal implementation. 

Then, as you are doing now, when we insert nodes at the end just take the size from meta. let me know what you think

imzhuhl commented:
> reduce memory usage from Uniform Value constant folder.

Do you mean to expect to reduce memory usage under both static and dynamic shapes?
You say “For non constructors, non-pointwise nodes, non-views return unknown_value” and “For constructors, always construct a tensor of size [1]”, I realize that these can reduce memory usage in both static and dynamic shape cases, so I have the above questions.

What I understand is to reduce memory usage by reducing the number of running nodes(early return) or the tensor size created by running nodes, is it accurate?

Also, I would like to ask if there is a way to quickly know which ops are constructors or views? (I currently know pointwise ops via torch.Tag.pointwise).

imzhuhl commented on torch/_inductor/config.py:
```diff
@@ -394,6 +394,9 @@ def fx_graph_remote_cache_default():
 joint_graph_constant_folding = True
 
+# support dynamic shape derivation in constant_fold_uniform_value
+enable_dynamic_constant_fold_uniform_value = False
```
ok

eellison commented:
> Do you mean to expect to reduce memory usage under both static and dynamic shapes?

Yes, this is to reduce memory overhead of constant folding for both cases.

> Also, I would like to ask if there is a way to quickly know which ops are constructors or views? (I currently know pointwise ops via torch.Tag.pointwise). 

There's not :( See https://github.com/pytorch/pytorch/issues/129020


But i think at this point, in the joint graph, all of the constructors will have been decomposed to `torch.ops.aten.full.default`. 

For views you can either use ops in partioners.py https://github.com/pytorch/pytorch/blob/main/torch/_functorch/partitioners.py#L1194-L1203 or check that the storage on input and output of op on the fake tensor meta is the same

imzhuhl committed:
[inductor] reduce memory usage for uniform value constant folding

Test plan:
```
python test/inductor/test_torchinductor_dynamic_shapes.py -k test_constant_fold_uniform_value_dynamic
python test/inductor/test_torchinductor_dynamic_shapes.py -k test_unwrap_storage_didnt_work_repro
```

imzhuhl commented:
hi @eellison, I am not clear with what constant_data_ptr_count does. Can you explain it? I have temporarily delete it.
I think this commit completes the idea of ​​reduce memory, please review it.

eellison reviewed (commented):
Looks great, thanks for making changes! 

Two things:

1 - don't think we are handling interaction of `x * sym_shape` right now

2 - Reuse `_SymHashingDict` to go from symint -> fx node

eellison commented:
cc @imzhuhl are you able to update this ? If not, I can take to finish line

imzhuhl commented:
> cc @imzhuhl are you able to update this ? If not, I can take to finish line

You can finish this, I haven't started yet in fact.

eellison committed:
Address comments

eellison committed:
Fix channels last error

eellison commented:
Replaced by https://github.com/pytorch/pytorch/pull/129686
